doctor/lib/text_extraction.py

```diff
@@ -0,0 +1,374 @@
+        .index.tolist()
+    )
+    blocks = [
+        filtered_data[filtered_data["block_num"] == block]
+        for block in block_ids
+    ]
+    return blocks
+
+
+def extract_with_ocr(page: pdfplumber.pdf.Page, strip_margin: bool) -> str:
+    """Extract the page using OCR
+
+    :param page:Pdf Plumber Page
+    :param strip_margin: If we should trim the margins
+    :return: The extracted content for the page
+    """
+
+    image = convert_pdf_page_to_image(page, strip_margin)
+    data = ocr_image_to_data(image)
+    content = ""
+    prev = {}
+    for words in data:
+        for index, word in words.iterrows():
+            content = insert_whitespace(content, word, prev)
+            content += get_word(word, image.size[0], strip_margin)
+            prev = word
+    content = cleanup_content(content, page.page_number)
+    return content
+
+
+def insert_whitespace(content: str, word: dict, prev: dict) -> str:
+    """Insert whitespace after or before word
+
+    :param content: The text extracted so far
+    :param word: The OCR extraction object
+    :param prev: The previous word object extracted
+    :return: The content with the whitespace appended
+    """
+    is_new_line = prev.get("line_num", 0) != word["line_num"]
+    is_new_par = prev.get("par_num", 0) != word["par_num"]
+    prev_end = prev.get("left", 1) + prev.get("width", 1)
+
+    # Add vertical whitespace
+    if is_new_line or is_new_par:
+        vertical_gap = word["top"] - (
+            prev.get("top", 0) + prev.get("height", 0)
+        )
+        content += "\n\n" if vertical_gap > 100 else "\n"
+        prev_end = 0
+
+    # add horizontal whitespace
+    content += " " * int(((word["left"] - prev_end) / 25))
+    return content
+
+
+def get_word(word_dict: dict, width: float, strip_margin: bool) -> str:
+    """Append word to content
+
+    This function determines if a word should be added to the page content
+    and adds the word.
+
+    tesseract provides confidence values for its OCR outputs.  We use those
+    confidence values to determine if something is a good OCR output, a
+    likely artifact and should be excluded or is bad ocr but not an artifact.
+
+    If a word has a zero confidence or starts on the left most edge of the paper
+    we return it as an empty string.  It is likely an artifact.
+
+    If a word has confidence below 40, a number that usually equates to 3 to 5
+    standard deviations from confidences found in other words is entirely in the
+    margin of the page - its likely an artifact as well.
+
+    If a confidence is below 5 - for a very short word - or for a very long word
+    its likely part of the document but we have no idea so we return a square
+    box to indicate that.  This is often caused by stamps or lines in case captions
+
+    Finally if a low confidence word starts in the right margin - its likely a
+    bad OCR that is multiple standard deviations away so we return the word as
+    empty squares.
```

Comment:
This is so very helpful, but again it should be interlaced with the code. It's also usually best practice not to put magic numbers in your comments (because the code changes and then the comments done, and, oh no, which is correct?).

Reply:
sounds good

Reply:
moved to be inline